lib/Transforms/Scalar/DxilLoopUnroll.cpp

```diff
@@ -0,0 +1,816 @@
+    return false;
+
+  // If the loop exit condition is not in the latch, then the loop is not rotated. Give up.
+  if (!cast<BranchInst>(Latch->getTerminator())->isConditional())
+    return false;
+
+  // Promote alloca's
+  if (!CleanedUpAllocas) {
+    CleanedUpAllocas = true;
+    Mem2Reg(*F, *DT, *AC);
+  }
+
+  SmallVector<BasicBlock *, 16> ExitBlocks;
+  L->getExitBlocks(ExitBlocks);
+  SetVector<BasicBlock *> ExitBlockSet(ExitBlocks.begin(), ExitBlocks.end());
+
+  SmallVector<BasicBlock *, 16> BlocksInLoop;
+  BlocksInLoop.append(L->getBlocks().begin(), L->getBlocks().end());
+  BlocksInLoop.append(ExitBlocks.begin(), ExitBlocks.end());
+
+  // Heuristically find blocks that likely need to be unrolled
+  SetVector<BasicBlock *> ProblemBlocks;
+  FindProblemBlocks(L->getHeader(), BlocksInLoop, ProblemBlocks);
+
+  // Keep track of the PHI nodes at the header.
+  SmallVector<PHINode *, 16> PHIs;
+  for (auto it = Header->begin(); it != Header->end(); it++) {
+    if (PHINode *PN = dyn_cast<PHINode>(it)) {
+      PHIs.push_back(PN);
+    }
+    else {
+      break;
+    }
+  }
+
+  SetVector<BasicBlock *> ToBeCloned; // List of blocks that will be cloned.
+  for (BasicBlock *BB : L->getBlocks())
+    ToBeCloned.insert(BB);
+
+  SetVector<BasicBlock *> NewExits; // New set of exit blocks as boundaries for LCSSA
+  SetVector<BasicBlock *> FakeExits; // Set of blocks created to allow cloning original exit blocks.
+  for (BasicBlock *BB : ExitBlocks) {
+    bool CloneThisExitBlock = ProblemBlocks.count(BB);
+
+    if (CloneThisExitBlock) {
+      ToBeCloned.insert(BB);
+
+      BasicBlock *FakeExit = BasicBlock::Create(BB->getContext(), "loop.exit.new");
+      F->getBasicBlockList().insert(BB, FakeExit);
+
+      TerminatorInst *OldTerm = BB->getTerminator();
+      OldTerm->removeFromParent();
+      FakeExit->getInstList().push_back(OldTerm);
+
+      BranchInst::Create(FakeExit, BB);
+      for (BasicBlock *Succ : successors(FakeExit)) {
+        for (Instruction &I : *Succ) {
+          if (PHINode *PN = dyn_cast<PHINode>(&I)) {
+            for (unsigned i = 0; i < PN->getNumIncomingValues(); i++) {
+              if (PN->getIncomingBlock(i) == BB)
+                PN->setIncomingBlock(i, FakeExit);
+            }
+          }
+        }
+      }
+
+      NewExits.insert(FakeExit);
+      FakeExits.insert(FakeExit);
+
+      // Update Dom tree
+      if (!DT->getNode(FakeExit))
+        DT->addNewBlock(FakeExit, BB);
+    }
+    else {
+      NewExits.insert(BB);
+    }
+  }
+
+  // Simplify the PHI nodes that have single incoming value. The original LCSSA form
+  // (if exists) does not necessarily work for our unroll because we may be unrolling
+  // from a different boundary.
+  for (BasicBlock *BB : BlocksInLoop)
+    SimplifyPHIs(BB);
+
+  // Re-establish LCSSA form to get ready for unrolling.
+  CreateLCSSA(ToBeCloned, NewExits, L, *DT, LI);
+
+  SmallVector<LoopIteration, 16> Clones;
+
+  const unsigned MaxIterations = 128;
```

Comment:
Can we make MaxIterations a pass argument with a default instead of an inline hard-coded const here?